core/src/main/scala/sttp/client/testing/SttpBackendStub.scala

```diff
@@ -211,3 +282,100 @@ object SttpBackendStub {
+    headers: Headers,
+    useHandler: Any => WS_RESULT
+) extends WhenOpenWebsocket[F, WS_HANDLER]
+
+private[testing] case class ReturnWebsocketResponse[F[_], WS_RESULT](headers: Headers, response: F[WS_RESULT])
+    extends WhenOpenWebsocket[F, NothingT]
+
+/**
+  * Simple stub for web sockets that uses queue of events for receive.
+  */
+class WebSocketStub[F[_], S](
+    events: mutable.Queue[F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]],
+    initialState: S,
+    makeNewEvents: (S, WebSocketFrame) => (S, List[F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]])
+)(implicit
+    val monad: MonadError[F]
+) extends WebSocket[F] {
+
+  private var state: S = initialState
+
+  override def isOpen: F[Boolean] = monad.unit(true)
+
+  override def receive: F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]] =
+    events.dequeueFirst(_ => true) match {
+      case Some(event) => event
+      case None        => monad.error(new Exception("Unexpected 'receive', no more prepared events."))
+    }
+
+  override def send(frame: WebSocketFrame, isContinuation: Boolean): F[Unit] =
+    monad.eval {
+      val (newState, newEvents) = makeNewEvents(state, frame)
+      newEvents.foreach(events.enqueue(_))
+      state = newState
+    }
+
+  /** Returns stub that has same prepared messages but adds new messages to receive in reaction to send.
+    *
+    * @param addReceived
+    */
+  def respondUsing(addReceived: WebSocketFrame => List[WebSocketFrame.Incoming]): WebSocketStub[F, Unit] =
+    respondUsingF(
+      addReceived.andThen(_.map(m => monad.unit(Right(m): Either[WebSocketEvent.Close, WebSocketFrame.Incoming])))
+    )
+
+  /** Returns stub that has same prepared messages but adds new messages to receive in reaction to send.
+    *
+    * @param addReceived
+    */
+  def respondUsingF(
+      addReceived: WebSocketFrame => List[F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]]
+  ): WebSocketStub[F, Unit] =
+    new WebSocketStub(
+      mutable.Queue.concat(events),
+      (),
+      (_, frame) => ((), addReceived(frame))
+    )
+
+  def respondUsingState[S2](initial: S2)(
+      onSend: (S2, WebSocketFrame) => (S2, List[WebSocketFrame.Incoming])
+  ): WebSocketStub[F, S2] =
+    respondUsingStateF(initial)((state, frame) => {
+      val (newState, messages) = onSend(state, frame)
+      (newState, messages.map(m => monad.unit(Right(m): Either[WebSocketEvent.Close, WebSocketFrame.Incoming])))
+    })
+
+  def respondUsingStateF[S2](initial: S2)(
+      onSend: (S2, WebSocketFrame) => (S2, List[F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]])
+  ): WebSocketStub[F, S2] = new WebSocketStub(mutable.Queue.concat(events), initial, onSend)
+
+}
+
+object WebSocketStub {
+
+  /** Creates stub that has given incoming frames prepared for 'receive' and doesn't add messages on 'send'.
+    * There is more powerful version that takes list of effects to return.
```

Comment:
Do you mean that this here is the more powerful version, or that somewhere else is a more powerful version?

Reply:
Somewhere else (postfix F)

Reply:
Which method is it?

Reply:
`withMessages` but I'm going to rename it to `withResponses`. I'm going one more round through this PR, focusing on docs.